### ISSUE TITLE
pdb2pqr runs not deterministic

### ISSUE THREAD

**Issue by tonigi** [Electrostatics/apbs-pdb2pqr#439]
Apparently pdb2pqr results depend on hash ordering, which is platform-dependent.  If you run 3PTB with -R (randomize hashes), HIS 91 turns out to be HID/HIE at random (roughly 50%).  This becomes an issue under Python 3, which has random hashes on by default.

To check:
```
for i in `seq 10`; do python2 -R main.py  --with-ph=7.0 --ff=parse 3PTB.pdb py2-$i.pqr; done
grep -c "HD1 HIS    91" py3-*.pqr
```

T

**Comment by keith923** [comment]
@kmonson, can you comment?  Do you know off the top of your head where this is happening?

**Comment by tonigi** [comment]
In [this branch](https://github.com/tonigi/apbs-pdb2pqr/tree/setuptools) I fixed it with liberal use of OrderedDicts (4bad162a6c03d79e109d6e9c9d5b69312bb59881).

**Comment by kmonson** [comment]
I'm not sure where the problem is in this case. We try to be deterministic everywhere it matters. I know that for debumping we use the same order. In the graphcut code in pdb2pka order doesn't matter if the algorithm is implemented correctly.

I could see a case (based on what changed in @tonigi's branch) where the hbond optimization could happen out of order and cause a problem. The algorithm checks each half of a potential bond and sometime the results of the first half causes the second half to be skipped. I was under the impression that it was symmetrical, but that may not be the case.

It would certainly explain the one pdb2pka test case we saw that was platform dependent.

**Comment by keith923** [comment]
It seems like this must have been a problem even before @tonigi ported the code to Python 3.  Right?  I mean, dict ordering has never been guaranteed.  At least not that I've seen.

**Comment by tonigi** [comment]
Yes. Python3 only made it easier to spot because it enforces randomization between runs (and not just between platforms).

**Comment by tonigi** [comment]
By the way, Py 3.6 again changed the behaviour of dictionaries, which are [now ordered](http://stackoverflow.com/questions/39980323/dictionaries-are-ordered-in-cpython-3-6) . Still, it is probably a good idea to be explicit in the use of OrderedDict if the goal is to be deterministic.

**Comment by sobolevnrm** [comment]
Thanks!

**Comment by sobolevnrm** [comment]
@intendo -- can you see if https://github.com/Electrostatics/apbs-pdb2pqr/commit/4bad162a6c03d79e109d6e9c9d5b69312bb59881 can be merged?
